bench/workgen/workgen.cpp

```diff
@@ -1328,37 +1443,84 @@ ThreadRunner::op_get_key_recno(Operation *op, uint64_t range, tint_t tint)
     return (rval % recno_count + 1); // recnos are one-based.
 }
 
+// Clear the table uri and tint value for a dynamic table operation.
+void
+ThreadRunner::op_clear_table(Operation *op)
+{
+    if (op->_random_table && op->is_table_op()) {
+        op->_table._uri = std::string();
+        op->_table._internal->_tint = 0;
+    }
+}
+
+// Generate and set a key and value for the operation managed by this thread.
+// Set the table uri and tint value for a dynamic table operation.
+void
+ThreadRunner::op_set_table(Operation *op, const std::string &uri, const tint_t tint)
+{
+    if (op->_random_table && op->is_table_op()) {
+        op->_table._uri = uri;
+        op->_table._internal->_tint = tint;
+    }
+}
+
+void
+ThreadRunner::op_kv_gen(Operation *op, const tint_t tint)
+{
+    if (!op->is_table_op()) {
+        return;
+    }
+
+    // A potential race: thread1 is inserting, and increments Context->_recno[] for fileX.wt.
+    // thread2 is doing one of remove/search/update and grabs the new value of Context->_recno[]
+    // for fileX.wt. thread2 randomly chooses the highest recno (which has not yet been inserted
+    // by thread1), and when it accesses the record will get WT_NOTFOUND. It should be somewhat
+    // rare (and most likely when the threads are first beginning). Any WT_NOTFOUND returns are
+    // allowed and get their own statistic bumped.
+    uint64_t recno = 0;
+    uint64_t range = op->_table.options.range;
+    if (op->_optype == Operation::OP_INSERT) {
+        if (op->_key._keytype == Key::KEYGEN_APPEND || op->_key._keytype == Key::KEYGEN_AUTO) {
+            if (op->_random_table) {
+                const std::shared_lock lock(*_icontext->_dyn_mutex);
+                recno = workgen_atomic_add64(&_icontext->_dyn_table_runtime.at(tint)._max_recno, 1);
+            } else {
+                recno = workgen_atomic_add64(&_icontext->_table_runtime[tint]._max_recno, 1);
+            }
+        } else {
+            recno = op_get_key_recno(op, range, tint);
+        }
+    } else {
+        recno = op_get_key_recno(op, range, tint);
+    }
+
+    VERBOSE(*this,
+      "OP " << op->_optype << " " << _icontext->_dyn_table_names[tint].c_str()
+            << ", recno=" << recno);
+
+    // Generate the key and value for the operation.
+    op->kv_gen(this, true, 100, recno, _keybuf);
+    if (OP_HAS_VALUE(op)) {
+        uint64_t compressibility =
+          (op->_table.options.random_value || _workload->options.random_table_values) ?
+          0 :
+          op->_table.options.value_compressibility;
+        op->kv_gen(this, false, compressibility, recno, _valuebuf);
+    }
+}
+
 int
-ThreadRunner::op_run(Operation *op)
+ThreadRunner::op_run_prepare(Operation *op)
```

Comment:
If I'm not mistaken, there is no transactional "prepare" implied by this function name?  If that's true, I'd suggest renaming it - I was looking for a prepare call.

Reply:
Fair point! I changed it to `op_run_setup` in [7dbaa36](https://github.com/wiredtiger/wiredtiger/pull/8798/commits/7dbaa36f79aee9ff5e042633644e2ce5e522c477) but open to suggestions for something better.

Reply:
`op_run_setup` sounds good.